### PR TITLE
[FEAT] Map stacktraces

### LINKED ISSUE #117: Emit and Consume Sourcemaps
sapper-error-trace

Currently when you have a bug in your code the server rendered code looks like the above.

It'd be very handy to emit and consume the sourcemaps so that stack trace gives us actionable info.

### PR DESCRIPTION
**Pull request by teoxoy**
closes #117

`sourcemap-stacktrace` takes in a stacktrace and uses sourcemaps (if it finds any) to map the paths to the sources

### CODE REVIEW

**Comment by Teoxoy** [comment]
@Conduitry has the issue been fixed in some other PR? Should I close this one?

If not, I can add the code from `sourcemap-stacktrace` package in sapper and delete the package.
I am not really up to maintaining it anymore, I don't know why I made it a package in the first place...

**Comment by benmccann** [comment]
No, I don't believe this issue has been fixed yet

**Comment by Teoxoy** [comment]
So, is there an issue with this PR? Do you guys want the code in sapper?

**Comment by benmccann** [comment]
I think mostly everyone's just been busy.  Also, errors weren't showing up at all due to https://github.com/sveltejs/sapper/pull/719, which made this one hard to review. Now that that's been fixed I think it's time to revisit this one.

>I can add the code from sourcemap-stacktrace package in sapper and delete the package.
I am not really up to maintaining it anymore, I don't know why I made it a package in the first place...

You said your package was based on [node-source-map-support](https://github.com/evanw/node-source-map-support), but exposes the needed functionality. Would you be able to file an issue with [node-source-map-support](https://github.com/evanw/node-source-map-support) to ask them to expose that functionality? Then we could consider using that package instead of having to add the code directly into Sapper

Also, I would suggest splitting out the linting fixes into a separate PR. We've been working on linting and you can see some of our progress in https://github.com/sveltejs/sapper/pull/1198, but unfortunately that doesn't fix the linting issues you found and we should investigate why. Perhaps we should setup prettier on the codebase

**Comment by Teoxoy** [comment]
I rechecked `node-source-map-support` and it has some node v8 stacktrace specifics + it doesn't look maintained to me.

I can remove the formatting changes and move the [src](https://github.com/Teoxoy/sourcemap-stacktrace/blob/835ea8be8f603b1d612be6f21e7ef527734e7148/src/index.ts) of `sourcemap-stacktrace` (95 LOC) in the repo.
Let me know if this is how we should move forward with this.

**Comment by antony** [comment]
How did this pass me by? This would be incredibly useful.

I would agree that replicating the source of that module here is probably desirable rather than bringing the whole thing in. Can we include a note so that it's clear that this is the case, and if the source module improves we can incorporate at a later date.

Also as @benmccann says, we don't accept linting fixes in PRs, unless that is the sole purpose of the PR, so they should be removed for this one.

Thanks!

**Comment by benmccann** [comment]
> it has some node v8 stacktrace specifics

V8 is the Javascript engine inside of node.js that parses and runs your Javascript. The same V8 engine is used inside of Chrome to run javascript in the Chrome browser. Google open-sourced the V8 engine and the builders of node.js used it to run Javascript in node.js. The current node.js binary cannot work without V8. So I'm not sure why it'd be an issue to have v8 specifics in the code?

If you're talking about Node 8.x instead, I'm also not sure it harms us at all if it supports Node 8.x

> it doesn't look maintained to me.

I'm not sure I'd call `source-map-support` unsupported. It does have a number of open issues and PRs, but also has had some commits regularly and was last updated only two months ago and has 20m weekly downloads.

Anyway, we don't have to block on getting the functionality implemented upstream. I wouldn't want to block this feature if we can't get the support we need. But we should at least ask for it before forking the project and copying the source here that way there's at least a possibility that we don't have to maintain a fork of the project long-term.

**Comment by Teoxoy** [comment]
I am aware of the v8/node relationship. What I meant to say is that `source-map-support` is using the [V8 stack trace API](https://v8.dev/docs/stack-trace-api) which doesn't work for us (this is part of the reason it was removed in #366).

I'm not sure why you refer to `source-map-support` as upstream since I mention in the readme of `sourcemap-stacktrace` that it is a reimplementation.

The relevant part to us is 1/6th of `source-map-support`; sourcemapping a stacktrace is quite simple, it's a matter of going trough each line of the stacktrace and trying to replace the path, line no. and column no.. Doing that by searching in the file that thew the error for a `sourceMappingURL` and mapping the orignal line and column numbers using [`source-map`](https://github.com/mozilla/source-map).

**Comment by benmccann** [comment]
> source-map-support is using the V8 stack trace API which doesn't work for us (this is part of the reason it was removed in #366).

Can you explain why it doesn't work for us? I don't see any reference to it in #366

**Comment by Teoxoy** [comment]
> notwithstanding that it doesn't currently apply to components

It has been a year since I worked on this - so, I don't actually remember why it didn't work.

**Comment by benmccann** [comment]
Ok. I'm fine with this PR if you're able to move the linting fixes to a new PR and rebase it

**Comment by Teoxoy** [comment]
Rebased, removed formatting/linting fixes and moved the src of `sourcemap-stacktrace` in the repo.

File: package.json [hunk 1]
```diff
@@ -21,6 +21,7 @@
     "html-minifier": "^4.0.0",
     "http-link-header": "^1.0.2",
     "shimport": "^1.0.1",
+    "source-map": "^0.7.3",
```
Comment (benmccann) [inline package.json:24]: we don't need this anymore since you're including the file directly, right?

Comment (Teoxoy) [reply package.json:24]: We do need that one. It's used to do the actual mapping.

**Comment by teoxoy** [comment]
I can't seem to get the test suite running in WSL and I currently don't have the time to debug it and also add tests.

**Comment by benmccann** [comment]
ok. thanks for all your work on this already. this lgtm

**Comment by pngwn** [comment]
This needs tests before it can be merged.

**Comment by benmccann** [comment]
I was just thinking about this some more... Does anyone know why we bundle the server? I wonder if it's necessary. If we didn't bundle it then I don't think we'd need source maps for the stack traces

**Comment by pngwn** [comment]
Bundling, even for the server, generally makes things a bit faster and smaller. It also makes things like building a docker image simpler as the server bundle contains all of the necessary code so you don't need to ship any dependencies.

**Comment by sdwvit** [comment]
please merge & release asap

**Comment by benmccann** [comment]
@sdwvit you can help it get merged. the main thing this needs is some tests. if you'd like to contribute we'd be happy to see this get merged much sooner

File: runtime/src/server/middleware/sourcemap_stacktrace.ts [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+
+                let rawSourceMap: RawSourceMap;
+                try {
+                    rawSourceMap = JSON.parse(srcMapData);
+                } catch {
+                    return input;
+                }
+
+                const consumer = new SourceMapConsumer(rawSourceMap);
+                const pos = consumer.originalPositionFor({
```
Comment (habibrosyad) [inline runtime/src/server/middleware/sourcemap_stacktrace.ts:74]: Apparently, with the latest version of `source-map`, the `consumer` here returns promise, therefore call to `originalPositionFor` will fail 😕

Comment (teoxoy) [reply runtime/src/server/middleware/sourcemap_stacktrace.ts:74]: You are right, it should be version 0.6.1

Comment (habibrosyad) [reply runtime/src/server/middleware/sourcemap_stacktrace.ts:74]: I'm planing to continue the works that you've done if you don't mind, especially adding tests for this...

Comment (teoxoy) [reply runtime/src/server/middleware/sourcemap_stacktrace.ts:74]: Not at all, I mentioned in one of my previous comments that I've been quite busy lately.

**Pull request closed without merging**